### ISSUE TITLE
请问，单GPU怎么解决训练过程中mae,mse不变的问题？

### ISSUE THREAD

**Issue by kexiaolong0121** [dk-liang/CLTR#30]
请问，单GPU怎么解决训练过程中mae,mse不变的问题？

**Comment by cnwxi** [comment]
同样的问题。

**Comment by cnwxi** [comment]
使用2*2080Ti，py3.6，torch1.8.0
![image](https://github.com/dk-liang/CLTR/assets/45331556/f0ef4cdf-60a2-4341-b37d-8e7bcf309985)
![image](https://github.com/dk-liang/CLTR/assets/45331556/cab666d6-41cc-4c2c-8993-a8bb3becca86)
学习率调整为`1e-5`
```SHELL
export OMP_NUM_THREADS=1
export MKL_NUM_THREADS=1 
python -m torch.distributed.launch --nproc_per_node=2 --master_port 8218 train_distributed.py --gpu_id '0,1' \                     
--gray_aug --gray_p 0.1 --scale_aug --scale_type 1 --scale_p 0.3 --epochs 1500 --lr_step 1200 --lr 1e-5 \ 
--batch_size 4 --num_patch 1 --threshold 0.35 --num_queries 700 \ 
--dataset nwpu --crop_size 256 --pre /media/ubuntu/2.0TB/wxy/CLTR/save_file/log_file/20240427_004616/checkpoint.pth --test_per_epoch 1  --test_patch --save
```
大概在几个epoch后测试的mae、mse值会发生变化，第一个epoch预测全为零，几个epoch后部分图片的预测开始变化。
建议先10个epoch后查看结果。

**Comment by kexiaolong0121** [comment]
![image](https://github.com/dk-liang/CLTR/assets/132921966/fa4e8c30-5bb6-4089-a916-acc02198dbcb)
我是除去了分布式训练的代码然后在jhu数据集上可以了，但在nwpu数据集上还是不行，哥可以加个联系方式交流一下吗

**Comment by cnwxi** [comment]
首页有邮箱

**Comment by kexiaolong0121** [comment]
调整一下学习率真的可以了，谢谢哥，你真厉害！

**Comment by cnwxi** [comment]
没事，互相帮助互相学习












未响应
***@***.***






---原始邮件---
发件人: ***@***.***&gt;
发送时间: 2024年4月27日(周六) 晚上8:17
收件人: ***@***.***&gt;;
抄送: ***@***.******@***.***&gt;;
主题: Re: [dk-liang/CLTR] 请问，单GPU怎么解决训练过程中mae,mse不变的问题？ (Issue #30)




 
调整一下学习率真的可以了，谢谢哥，你真厉害！

**Comment by IceIce1ce** [comment]
Hi @cnwxi, I also have a problem when training on my custom dataset with two GPUs. The MAE and MSE does not change after several test time. How can you solve this problem?

**Comment by cnwxi** [comment]
> Hi [@cnwxi](https://github.com/cnwxi), I also have a problem when training on my custom dataset with two GPUs. The MAE and MSE does not change after several test time. How can you solve this problem?

If you don't have enough graphics cards and memory to support the batch size settings like this, you should reduce the learning rate. This has been tested on the same dataset as the code. I'm not sure if it will work on custom datasets, but it's the easiest way to solve the problem of MAE/MSE not changing during training.

**Comment by IceIce1ce** [comment]
Hi, I have two A6000 GPUs and they can use batch size 16. However, the MAE and MSE still does not change. After changing the default learning rate from 1e-4 to 1e-5, the MAE and MSE decreased on the custom dataset. Therefore, I think the problem does not depend on the batch size. Anyway, thanks for your help.

**Comment by cnwxi** [comment]
> Hi, I have two A6000 GPUs and they can use batch size 16. However, the MAE and MSE still does not change. After changing the default learning rate from 1e-4 to 1e-5, the MAE and MSE decreased on the custom dataset. Therefore, I think the problem does not depend on the batch size. Anyway, thanks for your help.

I noticed you mentioned that the MAE and MSE did decrease after adjusting the learning rate, so modifying the learning rate actually worked, right? 
The batch size used in the original code is 16 and the number of graphics cards is 4, which actually in a single time the total batch size was 64. 
If you follow the default settings, your total batch size is 16*2. lr dose not match the batch size.
This is why you can't get the correct training effect when you use two graphics cards, the same batch size and the same learning rate. 

> PS: Since you have 2*A6000, maybe you should increase your batch size more.  

In general, the issue here lies in the matching of batch size and learning rate. You may need to try more combinations to achieve the optimal result. You can search Google for "learning rate batch size relation" to get more information.

**Comment by IceIce1ce** [comment]
Thanks for your suggestion.
